jb/fftw/time_delay_estimator_many.hpp

```diff
@@ -0,0 +1,172 @@
+ * @tparam T timeseries value type
+ * @tparam K timeseries dimensionality
+ */
+
+template <typename array_t, template <typename T, std::size_t K>
+                            class multi_array = jb::fftw::aligned_multi_array>
+class time_delay_estimator_many {
+public:
+  //@{
+
+  /**
+   * @name Type traits
+   */
+  /// The input timeseries type
+  using array_type = array_t;
+
+  /// The values stored in the input timeseries
+  using element_type =
+      typename jb::detail::array_traits<array_type>::element_type;
+
+  /// Extract T out of std::complex<T>, otherwise simply T.
+  using precision_type =
+      typename jb::extract_value_type<element_type>::precision;
+
+  /// The type used to store the DFT of the input timeseries
+  using frequency_array_type =
+      multi_array<std::complex<precision_type>, array_type::dimensionality>;
+
+  /// The type used to store the inverse of the DFT
+  using output_array_type =
+      multi_array<precision_type, array_type::dimensionality>;
+
+  /// The execution plan to apply the (forward) DFT
+  using dplan = jb::fftw::plan<array_type, frequency_array_type>;
+
+  /// The execution plan to apply the inverse (aka backward) DFT
+  using iplan = jb::fftw::plan<frequency_array_type, output_array_type>;
+
+  /// The type used to store the TDE confidence between two timeseries
+  using confidence_type = jb::fftw::tde_result<array_t, precision_type>;
+
+  /// The type used to store the TDE argmax between two timeseries
+  using argmax_type = jb::fftw::tde_result<array_t, std::size_t>;
+
+  /// The sqr sum of a timeseries
+  using sum2_type = jb::fftw::tde_result<array_t, precision_type>;
+
+  //@}
+
+  /**
+   * Constructs members using the multi array characteristic a and b.
+   *
+   * @param a multi array timeries
+   * @param b multi array timeries
+   *
+   */
+
+  time_delay_estimator_many(array_type const& a, array_type const& b)
+      : tmpa_{jb::detail::array_shape(a)}
+      , tmpb_{jb::detail::array_shape(b)}
+      , a2tmpa_{create_forward_plan(a, tmpa_, planning_flags())}
+      , b2tmpb_{create_forward_plan(b, tmpb_, planning_flags())}
+      , out_{jb::detail::array_shape(a)}
+      , tmpa2out_{create_backward_plan(tmpa_, out_, planning_flags())}
+      , nsamples_{jb::detail::nsamples(a)}
+      , num_timeseries_{jb::detail::element_count(a) / nsamples_} {
+    if (a.size() != b.size()) {
+      throw std::invalid_argument("size mismatch in time_delay_estimator ctor");
+    }
+  }
+
+  /**
+   * Compute the time-delay estimate between two timeseries a and b.
+   *
+   * @param a input timeseries
+   * @param b input timeseries
+   * @param sum2 contains sqr sum of one of the timeseries (a or b)
+   * @param confidence to return the confidence of the TDE(a,b)
+   * @param argmax to return argmax TDE(a,b)
+   */
+  void estimate_delay(
+      array_type const& a, array_type const& b, sum2_type const& sum2,
+      confidence_type& confidence, argmax_type& argmax) {
```

Comment:
I prefer output parameters to go first, that way they remind me of the old C library functions like memcpy()... think of the function call as an assignment, the outputs go to the left.